File: examples/clj/test/com/rpl/agent/basic/structured_langchain4j_agent_test.clj
(ns com.rpl.agent.basic.structured-langchain4j-agent-test
  (:require
   [clojure.test :refer [deftest testing is]]
   [com.rpl.agent-o-rama :as aor]
   [com.rpl.rama :as rama]
   [com.rpl.rama.test :as rtest]
   [com.rpl.agent.basic.structured-langchain4j-agent :refer [StructuredLangChain4jModule]]))

(deftest structured-langchain4j-agent-test
  (System/gc)
  (testing "StructuredLangChain4jAgent with real OpenAI model"
    (if (System/getenv "OPENAI_API_KEY")
      (with-open [ipc (rtest/create-ipc)]
        (rtest/launch-module! ipc StructuredLangChain4jModule {:tasks 1 :threads 1})

        (let [manager (aor/agent-manager ipc
                                         (rama/get-module-name
                                          StructuredLangChain4jModule))
              agent   (aor/agent-client manager "StructuredLangChain4jAgent")]

          (testing "returns structured response with all expected fields"
            (let [result (aor/agent-invoke agent "What is agent-o-rama?")]
              (is (map? result))
              (is (contains? result "question_type"))
              (is (contains? result "complexity"))
              (is (contains? result "main_topics"))
              (is (contains? result "answer"))
              (is (contains? result "confidence"))
              (is (string? (get result "answer")))
              (is (vector? (get result "main_topics")))))))
      (println "Skipping test: OPENAI_API_KEY not set"))))

File: examples/java/basic/src/test/java/com/rpl/agent/basic/StructuredLangchain4jAgentTest.java
package com.rpl.agent.basic;

import static org.junit.Assert.assertNotNull;

import com.rpl.agentorama.AgentClient;
import com.rpl.agentorama.AgentManager;
import com.rpl.rama.test.InProcessCluster;
import com.rpl.rama.test.LaunchConfig;
import org.junit.Test;

/**
 * Test class for StructuredLangchain4jAgent demonstrating structured output with LangChain4j.
 *
 * <p>This test demonstrates:
 *
 * <ul>
 *   <li>Structured output with JSON response format
 *   <li>OpenAI chat model integration
 *   <li>Question analysis with structured data
 * </ul>
 */
public class StructuredLangchain4jAgentTest {

  @Test
  public void testStructuredLangchain4jAgent() throws Exception {
    // Tests structured output from LangChain4j
    String apiKey = System.getenv("OPENAI_API_KEY");
    if (apiKey == null || apiKey.trim().isEmpty()) {
      System.out.println("Skipping test - OPENAI_API_KEY not set");
      return;
    }

    try (InProcessCluster ipc = InProcessCluster.create()) {
      // Deploy the agent module
      StructuredLangchain4jAgent.StructuredLangChain4jModule module =
          new StructuredLangchain4jAgent.StructuredLangChain4jModule();
      ipc.launchModule(module, new LaunchConfig(1, 1));

      // Get agent manager and client
      String moduleName = module.getModuleName();
      AgentManager manager = AgentManager.create(ipc, moduleName);
      AgentClient agent = manager.getAgentClient("StructuredLangChain4jAgent");

      // Test with a simple question
      String result = (String) agent.invoke("What is 2+2?");

      assertNotNull("Result should not be null", result);
    }
  }
}
